### ISSUE TITLE
Better file-box support: save bandwidth with URL type support via gRPC

### ISSUE THREAD

**Issue by huan** [wechaty/wechaty-puppet-hostie#115]
Today there's a discussion between @jcai and @su-chang and we found that we can not pass the MIME type via the gRPC stream service.

Then I realized that the stream has to transfer all the data between the server and client, which is not efficient for the URL type of the file box.

This needs to be improved, any discussion and suggestion are welcome.

## Update

1. the MIME type can be parsed from the filename

**Comment by su-chang** [comment]
### Hostie
Maybe we should add method `setMimeType` and `getMimeType` for `FileBoxChunk`.

```ts
async function packFileBoxToChunk (
  fileBox: FileBox,
): Promise<Readable<FileBoxChunk>> {
  const stream = new PassThrough({ objectMode: true })

  const chunk = new FileBoxChunk()
  chunk.setName(fileBox.name)
  
  // feature
  fileBox.ready()
  chunk.setMimeType(fileBox.mimeType)

  // FIXME: Huan(202010) write might return false
  stream.write(chunk)

  fileBox
    .pipe(encoder())
    .pipe(stream)

  return stream
}
```


```ts
async function unpackFileBoxFromChunk (
  stream: Readable<FileBoxChunk>,
): Promise<FileBox> {
  const chunk = await nextData(stream)
  if (!chunk.hasName()) {
    throw new Error('no name')
  }
  const fileName = chunk.getName()
  // feature
  const mimeType = chunk.getMimeType()

  const fileStream = new PassThrough({ objectMode: true })
  const transformedStream = stream.pipe(decoder())
  transformedStream.pipe(fileStream)
  stream.on('error', e => fileStream.emit('error', e))
  transformedStream.on('error', e => fileStream.emit('error', e))

  // feature
  const fileBox = FileBox.fromStream(fileStream, fileName, mimeType)

  return fileBox
}
```

### FileBox
```ts
public static fromStream (
  stream: Readable,
  name:   string,
  mimeType:   string,
): FileBox {
  const options: FileBoxOptions = {
    name,
    stream,
    type: FileBoxType.Stream,
  }
  this.mimeType = mimeType
  return new this(options)
}
```

**Comment by huan** [comment]
@su-chang Thanks for the thought and suggestion!

I have this idea at first, however, it seems we can get mime from the filename without any problem because the extension of the filename can be mapped to the mime type one by one.

See the mime npm module at <https://www.npmjs.com/package/mime>

**Comment by su-chang** [comment]
```ts
const url1 = 'https://s3.cn-north-1.amazonaws.com.cn/xiaoju-material/public/5cf7c197a0a2a410870e1582_1594369160229_null'
const url2 = 'http://www.sino-life.com/SL_LEM/grownService/showImgByFileId.do?fileId=2999300001-20201207080013-baa51d53d18a4e699310baf610e6b82a-000'

const name1 = '5cf7c197a0a2a410870e1582_1594369160229_null'
const name2 = 'showImgByFileId.do'
```

We can not get mime from the filename like these url.

https://github.com/broofa/mime/issues/235

**Comment by huan** [comment]
I believe that's not true.

You should be able to get almost all we need through the HTTP protocol.

See: 
https://github.com/huan/file-box/blob/0aa539acbd1e1100b688e5d47adc8ccafbe5f28c/src/file-box.ts#L339-L367

Please correct me if I'm not right.

**Comment by su-chang** [comment]
You are right, but `FileBox.ready()` is only work for `FileBox#URL`.

**Comment by huan** [comment]
So do you have any suggestions for improving FileBox so that it can work better?

I'd like to accept a PR to improve it if you have any.

**Comment by windmemory** [comment]
I also encountered an issue that sending videos across grpc with stream `FileBox` cause `Stream Removed` error today, so came into this issue and found the valuable discussion above. Here are my thoughts:

> 1. the MIME type can be parsed from the filename

@huan This is actually partly true. For those files with corresponding name and types, extracting MIME type from name works, but for those files that does not follow this rule, it does not work. And I believe this is why we also have the MIME type as a method in `FileBox`. So I would like to propose following solution:

### Add more fields into `FileBoxChunk`, current we only have `name` and `data`. We could add more like `type`, `url` and `mimeType`.

- These new fields are all optional
- For backward compatibility, when the type is not specified, the type will be default to stream
- We will pack and unpack the `FileBox` according to the type

**Comment by huan** [comment]
> For those files with corresponding name and types, extracting MIME type from name works, but for those files that does not follow this rule, it does not work

This case can be easily solved by set the name correctly before we save it to the FileBox, doesn't it?

**Comment by windmemory** [comment]
> This case can be easily solved by setting the name correctly before we save it to the FileBox, doesn't it?

There are some use cases that using an API endpoint to get a file, which is usually generated dynamically, like the `.do` example above, for these case, parsing mime type from name won't work.

> If I remember right, the reason that we have MIME type in the file box is because the URL type might have a URL without the filename with it.

I agree, there are some URL that does not have a name, and I believe this is a common use case. `FileBox` is parsing name from the url itself, and using the end of the url as the name, which is not accurate enough. With additional mime type, the information will be enough.

**Comment by huan** [comment]
TL;DR: MIME can be converted from a file name extension, and a file name extension can be converted to MIME too.

> like the .do example above, for these case, parsing mime type from name won't work.

That's the reason why I said `set the name correctly before we save it to the FileBox`. We should not set the `.do` to the file box name, but a valid file extension for the correct MIME type.
